kotlinx-coroutines-core/common/src/Annotations.kt

```diff
@@ -90,3 +90,28 @@ public annotation class ObsoleteCoroutinesApi
             "so stable API could be provided instead"
 )
 public annotation class InternalCoroutinesApi
+
+/**
+ * Marks declarations that cannot be safely inherited from.
+ */
+@Target(AnnotationTarget.CLASS)
+@RequiresOptIn(
+    level = RequiresOptIn.Level.WARNING, message =
+    "Inheriting from this kotlinx.coroutines API is unstable. " +
+        "Either new methods may be added in the future, which would break the inheritance, " +
+        "or correctly inheriting from it requires fulfilling contracts that may change in the future."
+)
+public annotation class ExperimentalForInheritanceCoroutinesApi
+
+/**
+ * Marks declarations that cannot be safely inherited from.
+ */
+@Target(AnnotationTarget.CLASS)
+@RequiresOptIn(
+    level = RequiresOptIn.Level.WARNING, message =
+    "This is a kotlinx.coroutines API that is not intended to be inherited from, " +
+        "as the library may handle predefined instances of this in a special manner. " +
+        "If you need to inherit from this, please describe your use case in " +
+        "https://github.com/Kotlin/kotlinx.coroutines/issues, so that we can provide a stable API for inheritance."
+)
+public annotation class InternalForInheritanceCoroutinesApi
```

Comment:
`InternalCoroutinesApi` has `ERROR` level - may be it makes sense to do the same here?

Reply:
It does, but I think we should delay making this an `ERROR`.

Imagine a situation: you're not trying to do anything fancy in your code, you just got confused and accidentally overrode something you shouldn't have, and now upgrading the library prevents you from using your code that you know used to work. If you got confused, it means you don't understand that part of your codebase all that well and will need to allocate some time to fixing it. Warnings are more gentle.